### PR TITLE
Ssh cert feature

### PR DESCRIPTION
**Pull request by angelmoo**
added commands:
**sshenroll** for requesting a certificate.
**sshpickup** for  retrieving a certificate.
added test cases for testing the new functionalities.

### CODE REVIEW

File: cmd/vcert/commands.go [hunk 1]
```diff
@@ -145,6 +147,24 @@ var (
...
 		vcert getpolicy -k <VaaS API key> -z "<app name>\<CIT alias>"`,
 	}
+
+	commandSshPickup = &cli.Command{
+		Before:    runBeforeCommand,
+		Name:      commandSshPickupName,
+		Flags:     sshPickupFlags,
+		Action:    doCommandSshPickup,
+		Usage:     "To retrieve a SSH Certificate",
+		UsageText: `vcert sshpickup -u https://tpp.example.com -t <TPP access token> --pickup-id <ssh cert DN>"`,
```
Comment (tr1ck3r) [inline cmd/vcert/commands.go:157]: Trailing double-quote should be removed.

File: cmd/vcert/commands.go [hunk 2]
```diff
@@ -145,6 +147,24 @@ var (
...
+	}
+
+	commandSshEnroll = &cli.Command{
+		Before:    runBeforeCommand,
+		Name:      commandSshEnrollName,
+		Flags:     sshEnrollFlags,
+		Action:    doCommandEnrollSshCert,
+		Usage:     "To enroll a SSH Certificate",
+		UsageText: `vcert sshenroll -u https://tpp.example.com -t <TPP access token> --template "<val>" --id <val> --principal "bob" --principal "alice" --valid-hours 1"`,
```
Comment (tr1ck3r) [inline cmd/vcert/commands.go:166]: Trailing double-quote should be removed.

File: cmd/vcert/commands.go [hunk 3]
```diff
@@ -397,11 +592,22 @@ func doCommandCredMgmt1(c *cli.Context) error {
...
 				User:     cfg.Credentials.User,
 				Password: cfg.Credentials.Password,
 				Scope:    flags.scope,
-				ClientId: flags.clientId})
+				ClientId: flags.clientId}
+
+			if flags.sshCred {
+				auth.Scope = "ssh:manage"
+			} else if flags.pmCred {
+				auth.Scope = "configuration:manage"
```
Comment (tr1ck3r) [inline cmd/vcert/commands.go:605]: @angelmoo I think this was my fault for telling you wrong... I think policy management requires `"certificate:manage"` for `getpolicy` and `"configuration:manage"` for `setpolicy`, right?  If so, we should get rid of the `--cpm` flag and update the scope for the `--pm` flag to `"certificate:manage,revoke;configuration:manage"`.

File: cmd/vcert/validators.go [hunk 4]
```diff
@@ -691,3 +693,86 @@ func validateSetPolicyFlags(commandName string) error {
...
+
+func validateSshEnrollFlags(commandName string) error {
+
+	err := validateConnectionFlags(commandName)
+	if err != nil {
+		return err
+	}
+
+	if flags.sshCertKeyId == "" {
+		return fmt.Errorf("a ssh certificate key id is requiered")
```
Comment (tr1ck3r) [inline cmd/vcert/validators.go:705]: "An ID is required for the SSH certificate (--id)"

File: cmd/vcert/validators.go [hunk 5]
```diff
@@ -691,3 +693,86 @@ func validateSetPolicyFlags(commandName string) error {
...
+	}
+
+	err = validateExistingFile(flags.sshCertKeyId)
+	if err != nil {
+		return err
+	}
+
+	if flags.sshCertTemplate == "" {
+		return fmt.Errorf("ca value is required")
```
Comment (tr1ck3r) [inline cmd/vcert/validators.go:714]: "CA value is required (--ca)"

File: cmd/vcert/commands.go (hunk 1 above)
Comment (angelmoo) [reply cmd/vcert/commands.go]: fixed on most recent commit

File: cmd/vcert/commands.go (hunk 2 above)
Comment (angelmoo) [reply cmd/vcert/commands.go]: fixed on most recent commit

File: cmd/vcert/validators.go (hunk 4 above)
Comment (angelmoo) [reply cmd/vcert/validators.go]: fixed on most recent commit.

File: cmd/vcert/commands.go (hunk 3 above)
Comment (angelmoo) [reply cmd/vcert/commands.go]: yes, and fixed in most recent commit

File: cmd/vcert/validators.go (hunk 5 above)
Comment (angelmoo) [reply cmd/vcert/validators.go]: as we changed the value to be template then I wrote something like this: "certificate issuing template value is required (--template)" please let me know if you agree with this or you have some changes on that.

File: cmd/vcert/flags.go [hunk 6]
```diff
@@ -517,6 +529,101 @@ var (
...
+
+	flagSshCertGuid = &cli.StringFlag{
+		Name:        "guid",
+		Usage:       "A value that uniquely identifies the certificate request.",
+		Destination: &flags.sshCertGuid,
+	}
+
+	flagSshPrivKeyPassphrase = &cli.StringFlag{
+		Name: "pk-passphrase",
+		Usage: "The passphrase which will be used to wrap the generated private key before it is returned in the API response." +
```
Comment (tr1ck3r) [inline cmd/vcert/flags.go:612]: Why can't we use `key-passphrase` for this use case?

File: cmd/vcert/flags.go [hunk 7]
```diff
@@ -517,6 +529,101 @@ var (
...
+			" This is applicable only in case of service-generated keypair.",
+		Destination: &flags.sshCertGuid,
+	}
+
+	flagSshCertExtension = &cli.StringSliceFlag{
+		Name:  "extensions",
```
Comment (tr1ck3r) [inline cmd/vcert/flags.go:618]: Should this be `extension` since you'd specify it multiple times if you need to add more than one extension?

**Review by tr1ck3r — APPROVED** [review]
Looks good @angelmoo I had a couple more questions.

File: cmd/vcert/flags.go (hunk 7 above)
Comment (angelmoo) [reply cmd/vcert/flags.go]: Changed on most recent commit.

File: cmd/vcert/flags.go (hunk 6 above)
Comment (angelmoo) [reply cmd/vcert/flags.go]: yes actually this is duplicated, key-passphrase is the one that we use

**Review by marcos-albornoz — APPROVED** [review]

**Review by ricrodriguezg — APPROVED** [review]

**Review by rvelaVenafi — APPROVED** [review]

**Pull request merged**